### PR TITLE
[feat][Ascend][Mindspore]: support model-impl of mindspore

### PR DESCRIPTION
**Pull request by chz34**
## Motivation

Following the plan of: #8199 

To support MindSpore on Ascend NPU as an alternative option of model execution. 

Use `model-impl` to selelct **MindSpore** implemented model for inference. MindSpore could provide graph based fusion and execution which have better performance on Ascend, especially for small batch.
 
This is the first try to merge the model execution of MindSpore and Pytorch. So it could also be a potential way to support other framework.

## Modifications

Generally, add the frist support model of mindspore implemented model Qwen3 dense model, and add the support to convert pytorch data and communication to other framework.

1. Add an option to the server_args `model-impl`
2. Add a `mindspore_runner` to set up some environment of MindSpore, and reuse the collective communication group of pytorch on Ascend. For now there is not a real Runner, only some preparation. 
3. Add a mindspore model entry as the interface between pytorch and mindspore models, while using `--model-impl mindspore`.  It will create the exact model accroding to the config.
4. mindspore model entry will registry the real mindspore model implementations from an individual package `sgl-mindspore`, which will be from another repo.

**SGLang keeps running normally in environments where MindSpore is not installed,** just `--model-impl mindspore` is not available.

## Accuracy Tests

Accuracy test of Qwen3-8B using gsm8k: 0.908

ScreenShot_20250815162057

## Benchmarking and Profiling

Benchmarking for single device of Qwen3-8B

``` shell
#Input tokens: 33839
#Output tokens: 21640
Starting warmup with 1 sequences...
Warmup completed with 1 sequences. Starting main benchmark run...
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100/100 [00:49<00:00,  2.04it/s]

============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    inf
Max request concurrency:                 not set
Successful requests:                     100
Benchmark duration (s):                  49.03
Total input tokens:                      33839
Total generated tokens:                  21640
Total generated tokens (retokenized):    21640
Request throughput (req/s):              2.04
Input token throughput (tok/s):          690.15
Output token throughput (tok/s):         441.35
Total token throughput (tok/s):          1131.49
Concurrency:                             23.68
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   11610.62
Median E2E Latency (ms):                 9387.14
---------------Time to First Token----------------
Mean TTFT (ms):                          1980.17
Median TTFT (ms):                        2044.41
P99 TTFT (ms):                           2731.71
---------------Inter-Token Latency----------------
Mean ITL (ms):                           44.71
Median ITL (ms):                         38.14
P95 ITL (ms):                            77.80
P99 ITL (ms):                            83.30
Max ITL (ms):                            2206.82
==================================================
==

```

Benchmark for 4 devices of Qwen3-32B of tp=4
```shell
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    inf
Max request concurrency:                 not set
Successful requests:                     100
Benchmark duration (s):                  80.01
Total input tokens:                      33839
Total generated tokens:                  21640
Total generated tokens (retokenized):    21640
Request throughput (req/s):              1.25
Input token throughput (tok/s):          422.94
Output token throughput (tok/s):         270.47
Total token throughput (tok/s):          693.40
Concurrency:                             23.39
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   18711.13
Median E2E Latency (ms):                 16017.74
---------------Time to First Token----------------
Mean TTFT (ms):                          3330.45
Median TTFT (ms):                        3480.48
P99 TTFT (ms):                           4614.00
---------------Inter-Token Latency----------------
Mean ITL (ms):                           71.41
Median ITL (ms):                         56.83
P95 ITL (ms):                            126.90
P99 ITL (ms):                            150.89
Max ITL (ms):                            4462.20
==================================================

```

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/models/mindspore.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+
+        logger.info(
+            "MindSporeForCausalLM tp size %d tp rank %d",
+            get_tensor_model_parallel_world_size(),
+            get_tensor_model_parallel_rank(),
+        )
+
+        model_type = self.config.model_type
+        if model_type not in type_model_map:
+            raise ValueError(f"Unsupported arch {arch}")
```
[the remark this answers is not in the document]
Comment (chz34) [reply python/sglang/srt/models/mindspore.py]: Fixed.

File: python/sglang/srt/models/mindspore_models/qwen3.py [hunk 2]
```diff
@@ -0,0 +1,973 @@
...
+        self,
+        config: Qwen3Config,
+        quant_config: Optional[QuantizationConfig] = None,
+        prefix: str = "",
+    ) -> None:
+        super().__init__()
+        self.prev_prefill = False
+
+        self.config = config
+        setattr(self.config, "param_dtype", dtype.bfloat16)
```
[the remark this answers is not in the document]
Comment (chz34) [reply python/sglang/srt/models/mindspore_models/qwen3.py]: Get from config.

File: python/sglang/srt/models/mindspore.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+            return mutable(self.key_cache), mutable(self.value_cache)
+
+        for i in range(self.config.num_hidden_layers):
+            k_cache = forward_batch.token_to_kv_pool.get_key_buffer(i)
+            v_cache = forward_batch.token_to_kv_pool.get_value_buffer(i)
+
+            self.key_cache.append(tensor_torch2ms(k_cache))
+            self.value_cache.append(tensor_torch2ms(v_cache))
+
+        return mutable(self.key_cache), mutable(self.value_cache)
```
[the remark this answers is not in the document]
Comment (chz34) [reply python/sglang/srt/models/mindspore.py]: `token_to_kv_pool` is created only once for each model. And here is the cache for each model instance. 

Using the cache to avoid redundant conversion from pytorch tensor to mindspore ones, which will spend more time.

File: python/sglang/srt/models/mindspore.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        model_inputs = self.prepare_inputs(input_ids, positions, forward_batch)
+        # prepare model inputs
+        model_inputs = self.model.prepare_inputs(forward_batch, model_inputs)
+
+        logits = self.model(**model_inputs)
+
+        # TODO: npu tensor ms2torch error to be fix, remain issues of torch_npu to get tensor from dlpack
+        logits_result = LogitsProcessorOutput(
+            next_token_logits=torch.Tensor(logits.asnumpy()).to(input_ids.device)
+        )
```
[the remark this answers is not in the document]
Comment (chz34) [reply python/sglang/srt/models/mindspore.py]: Create tensor by `from_dlpack` still have some issues in PyTorch with NPU. So currently can only use `asnumpy()` instead.

The logits of small batch is not a really big size of data, it may impact the performance a bit, but for now it's ok.

Waiting for the issues of Pytorch npu `from_dlpack` to be solved.

File: python/sglang/srt/model_executor/mindspore_runner.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+def init_ms_distributed(world_size, rank, local_rank, server_args, port):
+    if server_args.dist_init_addr:
+        master_addr, master_port = _get_host_and_ip(server_args.dist_init_addr)
+        dist_init_method = f"tcp://{master_addr}:{master_port + 33}"
+    else:
+        dist_init_method = f"tcp://{server_args.host}:{port + 33}"
+    set_ms_parallel_env(rank, local_rank, world_size, dist_init_method)
+
+    ms.set_context(infer_boost="on", jit_level="O0")
```
[the remark this answers is not in the document]
Comment (chz34) [reply python/sglang/srt/model_executor/mindspore_runner.py]: "O0" is enough for infernece, with better control of the inference. Higher optimization level is usually used for training, with more complex computing graph.

File: docs/references/mindspore_models.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    --model-impl mindspore \
+    --attention-backend ascend \
+    --log-level DEBUG
+```
+
+Enable mindspore info and debug logging by setting environments.
+
+```bash
+export GLOG_v=1  # INFO
+export GLOG_v=0  # DEBUG
```
[the remark this answers is not in the document]
Comment (chz34) [reply docs/references/mindspore_models.md]: 0 for Debug, 1 for Info, confirmed.

File: python/sglang/srt/model_executor/mindspore_runner.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+def _get_host_and_ip(distributed_init_method):
+    try:
+        ip_str, port_str = distributed_init_method.split(":")[-2:]
+        ip = ip_str.split("/")[-1]
+        port = int(port_str)
+    except Exception as e:
+        raise RuntimeError(
+            "Cannot get host and port information from %s, error: %s!"
+            % (distributed_init_method, str(e))
+        )
```
[the remark this answers is not in the document]
Comment (chz34) [reply python/sglang/srt/model_executor/mindspore_runner.py]: Need to catch all kinds of exception and raise the runtime error to show the corresponding information.

File: python/sglang/srt/models/mindspore_models/qwen3.py [hunk 8]
```diff
@@ -0,0 +1,973 @@
...
+        if is_prefill:
+            self.model.phase = "prefill"
+        else:
+            self.model.phase = "increment"
+
+        hidden_state = self.model(**model_inputs)
+
+        # TODO: In pure decode scenarios, cumsum and gather operations will be redundant .
+        q_seq_lens = mint.cumsum(q_seq_lens, 0)
+        hidden_state = self.gather(hidden_state, q_seq_lens - 1, 0)
```
[the remark this answers is not in the document]
Comment (chz34) [reply python/sglang/srt/models/mindspore_models/qwen3.py]: The condition also leads to bad performance in some situation. 
Whether use a condition or not is more complex, so now keep this just to make the logits more simple.

File: docs/references/mindspore_models.md [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+cd sgl-mindspore
+pip install -e .
+```
+
+You will need to install the following packages, due to the support of tensor conversion through `dlpack` on 3rd devices, the minimum version of  `PyTorch` is 2.7.1
+
+```shell
+pip install mindspore
+pip install torch==2.7.1
+pip install torch_npu==2.7.1rc1
```
Comment (iforgetmyname) [inline docs/references/mindspore_models.md]: maybe update to the latest release of `torch_npu`?

btw we are planning to upgrade to torch 2.8.0, could you please test if 2.8.0 is supported?

**Review by ping1jing2 — APPROVED** [review]

File: python/sglang/srt/models/mindspore.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the SGLang project
+import logging
+from typing import Any, Iterable, Optional, Tuple
+
+import mindspore as ms
+import numpy as np
+import torch
+import torch_npu
+from mindspore import Tensor, mint, mutable
```
Comment (iforgetmyname) [inline python/sglang/srt/models/mindspore.py]: wrap these up with `_is_npu` and try catch otherwise it will give annoying `import error` warnings

**Comment by iforgetmyname** [comment]
Hi, @chz34 pls check ci failure: https://github.com/sgl-project/sglang/actions/runs/19429030716/job/55583275939?pr=9234

File: docs/references/mindspore_models.md (hunk 9 above)
Comment (wangtiance) [reply docs/references/mindspore_models.md]: We added dlpack API to torch_npu 2.7.1 and above, so any version >= 2.7.1 should work. We can test 2.8.0

File: python/sglang/srt/models/mindspore.py (hunk 10 above)
Comment (wangtiance) [reply python/sglang/srt/models/mindspore.py]: okay

File: python/sglang/srt/model_executor/mindspore_runner.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+import multiprocessing as mp
+import os
+import sys
+from pathlib import Path
+
+import mindspore as ms
+import torch
+from mindspore._c_expression import GroupOptions
+from mindspore.communication import create_group
```
Comment (ping1jing2) [inline python/sglang/srt/model_executor/mindspore_runner.py]: use `if is_npu:`

File: python/sglang/srt/models/mindspore.py [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+)
+from sglang.srt.layers.logits_processor import LogitsProcessorOutput
+from sglang.srt.layers.quantization.base_config import QuantizationConfig
+from sglang.srt.model_executor.forward_batch_info import ForwardBatch
+from sglang.srt.models.registry import import_model_classes
+from sglang.srt.utils import is_npu
+
+_is_npu = is_npu()
+
+if is_npu:
```
Comment (ping1jing2) [inline python/sglang/srt/models/mindspore.py]: wrong, `_is_npu`

File: docs/references/mindspore_models.md (hunk 9 above)
Comment (chz34) [reply docs/references/mindspore_models.md]: used ">=" instead.

File: python/sglang/srt/model_executor/mindspore_runner.py (hunk 11 above)
Comment (chz34) [reply python/sglang/srt/model_executor/mindspore_runner.py]: added.

File: python/sglang/srt/models/mindspore.py (hunk 12 above)
Comment (chz34) [reply python/sglang/srt/models/mindspore.py]: Typo, fixed.

File: docs/references/mindspore_models.md (hunk 9 above)
Comment (iforgetmyname) [reply docs/references/mindspore_models.md]: sounds good, if users can reuse our released docker images and give them a little tweaks to make mindspore work, that will be much much better

File: python/sglang/srt/model_executor/mindspore_runner.py (hunk 11 above)
Comment (iforgetmyname) [reply python/sglang/srt/model_executor/mindspore_runner.py]: don't, import of this whole file is guarded by `_is_npu` here
https://github.com/sgl-project/sglang/pull/9234/files#diff-65c6ac2c41977f68e460f18e35053b97089631f88a9958b0796343fccee78a67R372-R373

try to avoid `_is_npu` as much as possible to reduce potential overhead

File: test/srt/run_suite.py [hunk 13]
```diff
@@ -371,6 +371,7 @@
         TestFile("test_wave_attention_backend.py"),
         TestFile("test_weight_version.py"),
         TestFile("test_deepseek_v32_cp_single_node.py", 275),
+        TestFile("test_mindspore_models.py"),
```
Comment (iforgetmyname) [inline test/srt/run_suite.py]: create a `__not_in_ascend_ci__` section under `suite_ascend` and move it to there

File: test/srt/ascend/test_mindspore_models.py
Comment (iforgetmyname) [inline test/srt/ascend/test_mindspore_models.py:1]: move this test file to `srt/ascend/`

File: test/srt/run_suite.py (hunk 13 above)
Comment (chz34) [reply test/srt/run_suite.py]: moved.

File: test/srt/ascend/test_mindspore_models.py
Comment (chz34) [reply test/srt/ascend/test_mindspore_models.py:1]: moved.

File: python/sglang/srt/model_executor/mindspore_runner.py (hunk 11 above)
Comment (chz34) [reply python/sglang/srt/model_executor/mindspore_runner.py]: Now keeping the current state.  The whole runner file will only be dynamic imported if model_impl == mindspore and is_npu

**Review by iforgetmyname — APPROVED** [review]
looks good, thx for your contribution
